Add fetch hook into tracer.

dristic opened the pull request:
This is a first draft on how to hook static string data. This would allow you to build the string source of a file yourself, then pass it into the builder to resolve dependencies. Example:

```javascript
builder.build('foo.js', 'bar.js', {
  fetch: function (data) {
    if (data.name.indexOf('foo.js') !== -1) { return 'var bar = 2;'; }
  }
});
```

Fixes #330

guybedford commented on test/test-build.js:
```diff
@@ -39,6 +39,18 @@ function doTests(transpiler) {
+        if (data.name.indexOf('foo.js') !== -1) {
+          return fs.readFileSync('test/fixtures/test-tree/amd-1.js', 'utf8');
+        }
+      }
```
Nice API! How about injecting this into the loader itself by implementing in the `lib/builder.js` reset() function - `var loaderFetch = builder.loader.fetch; builder.loader.fetch = function(load) { return (builder.fetch || loaderFetch).call(this, load); }` and then having this config option set `builder.fetch`?

dristic replied:
Thanks! Do you think it is good practice to override the loader's functions like that? It might make it less predictable what the fetch function does in other areas of the framework, making it less modular.

What do you think about adding a fallthrough case as well?
```javascript
function (load) {
  var ret = builder.fetch(load);
  if (!ret) ret = loaderFetch(load);
  return ret;
}
```

This way it will fall through to using the default loader if you do not return anything from your `fetch` function.

guybedford replied:
The suggestion above is based on a single extension of the loader that checks for a `builder.fetch` instance function on the builder not the loader itself. That is, setting `builder.fetch = null` completely reverses the fetch hook override. I'd prefer not to provide a fallthrough unless there is a good use case to justify it?

dristic replied:
My only thought was being able to provide a complex expression or a glob, then provide the source as a string to a select few files. I could not be entirely understanding the system though.

I will look into the suggestion a bit more today and try to update this PR :smile:

dristic commented:
Okay I refactored it a bit to use the fetch override. One of the issues is that since we are completely overriding it, you can't pass in a string and have systemjs build the deps without the pass function. This kind of defeats the purpose so I added in a `pass()` that lets you fall back to the default implementation.

guybedford commented:
Thanks, this looks great. Agreed about the pass through concept - perhaps we could use the same pattern as we do for plugins then - https://github.com/systemjs/systemjs/blob/master/docs/creating-plugins.md#sample-css-plugin where the original `fetch` becomes the argument?

dristic commented:
Ah okay, I did not see this! I was modeling it after the `beforeEach(function (done) { done(); })` pattern from jasmine. I will update the PR and also add to the readme. :+1:

guybedford commented:
Great, thanks, just let me know when its ready as new commits don't send notifications.

dristic commented:
Sorry for the delay on this. Updated my code and the README as well!

guybedford commented on lib/trace.js:
```diff
@@ -35,7 +35,7 @@ Trace.prototype.traceModule = function(moduleName, traceAllConditionals, conditi
   .then(function(canonicalConditionalEnv) {
     if (!conditionsOnly)
-      return self.getAllLoadRecords(moduleName, traceAllConditionals, canonicalConditionalEnv);
+      return self.getAllLoadRecords(moduleName, traceAllConditionals, canonicalConditionalEnv, undefined);
```
This shouldn't be needed?

guybedford commented on lib/arithmetic.js:
```diff
@@ -136,7 +136,7 @@ exports.traceExpression = function(builder, expression, traceOpts) {
         
         // tree . tree
-        return builder.tracer.traceModule(op.moduleName, traceOpts.traceAllConditionals, traceOpts.conditions, traceOpts.traceConditionsOnly)
+        return builder.tracer.traceModule(op.moduleName, traceOpts.traceAllConditionals, traceOpts.conditions, traceOpts.traceConditionsOnly, traceOpts.fetch)
```
This shouldn't be needed either?

guybedford commented on lib/builder.js:
```diff
@@ -409,6 +409,25 @@ Builder.prototype.bundle = function(expressionOrTree, outFile, opts) {
+      var args = Array.prototype.slice.call(arguments, 0);
+      args.push(loaderFetch.bind(self.loader));
+
+      var source = (self.fetch || loaderFetch).apply(self.loader, args);
```
I think it's fine to use `(self.fetch || loaderFetch).call(self.loader, load, loaderFetch)` which cuts out the above two lines.

dristic replied:
I swear I had a good reason for using apply :smile: Will fix

guybedford commented on lib/builder.js:
```diff
@@ -409,6 +409,25 @@ Builder.prototype.bundle = function(expressionOrTree, outFile, opts) {
+      if (typeof source === 'undefined') {
+        console.warn('Provided fetch function but did not return source for', load.name);
+        source = '';
+      }
```
:+1:

guybedford commented:
Thanks for following up here!

guybedford commented on README.md:
````diff
@@ -187,6 +187,28 @@ builder.config({
 });
 ```
 
+### Overriding Fetch
````
Can we move this down to the bottom of the docs, as it is an advanced feature?

dristic commented:
Oops, must have had some leftover changes from my previous method of doing this. I will update the branch with the fixes requested.
